### DISCUSSION TITLE
Is it possible to play a certain Soundfile depending on Events in my Program?

### DISCUSSION THREAD

**Discussion by pdimitri** [pschatzmann/arduino-audio-tools#882] [Q&A] [answered]
I've read all your Blog Entires about mp3 Streaming and File Players, but yet - even in Code Snippets from other People here in the Discussions Section - it's always so that all Files from a Directory (SD-FAT or SPIFFS or an Array of URLs) are all grabbed at the same Time and played once one after another during the Program Run.

I haven't found a single Example on how to (Pre-) Load a mp3 File, and play it once a certain Condition in my Program is met. Ideally with very low Latency. I am programming some Kind of Game with the ESP32, where I want to give the Players audible Feedback from the Game, such as "you won" or "you lost".  I don't need final Example Code, but only some Hints if what I want to achieve is possible with the Library, and where I may get more INformations.

**Comment by pschatzmann** [comment, marked as answer]
I don't really understand where you see any issue: It should be pretty simple to change the [example to play one file](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-audiokit/streams-sd-audiokit/streams-sd-audiokit.ino) to playing any file triggered by any event.  To keep things simple you could try to use copier.copyAll() instread of running individual copier.copy() steps in the loop.
What have you tried and where did you fail ?

**Reply by pdimitri** [reply]
Well, I'm getting completely lost with all the Audio-Sources and Sinks in your Frameworks, and don't really understand the Differences between some of the Stream- and Audio-Source Classes. I want to read an mp3-File from the SPIFFS Storage, and send the resulting Audio-Stream via Bluetooth to a Bluetooth-Speaker. 

My first Problem is, that in all SPIFFS Examples an AudioSourceSPIFFS Object is used, and the only Constructor for this is:

AudioSourceSPIFFS(const char *startFilePath = "/", const char *ext = ".mp3")

Thats why I asked if it would be also possible to load a certain File instead of a List of all mp3 Files in the Filesystem.

OK, I could also load the File directly from the File System instead, but what should I do with the File to stream it as an A2DPStream? 

Currently I have this Code:

`#include <Arduino.h>
#include <Wire.h>
#include "AudioTools.h"
#include "AudioLibs/AudioA2DP.h"
#include "AudioLibs/AudioSourceSPIFFS.h"
#include "AudioCodecs/CodecMP3Helix.h"


A2DPStream out;
EncodedAudioStream decoder(new MP3DecoderHelix());
StreamCopy copier; 
File audioFile;

void printMetaData(MetaDataType type, const char* str, int len){
  Serial.print("==> ");
  Serial.print(toStr(type));
  Serial.print(": ");
  Serial.println(str);
}

void setup() {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);

  if(!SPIFFS.begin(true)){
        Serial.println("An Error has occurred while mounting SPIFFS");
        return;
   }

  File audioFile = SPIFFS.open("/002.mp3", FILE_READ);
    if(!audioFile){
        Serial.println("Failed to open MP3 file for reading");
        return;
    } else {
       Serial.println("Successfully loaded mp3 File");
    }
  
  auto cfg = out.defaultConfig(TX_MODE);
  cfg.name = "JBL GO 2";

  out.begin(cfg);
 
  decoder.begin();
 
  copier.setCheckAvailableForWrite(false);
  copier.begin(decoder, audioFile); 
}

void loop() {
  if (!copier.copy()) {
    stop();
  }
}`

The Part after setting up the Bletooth Connection is identical to the [example to play one File](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-audiokit/streams-sd-audiokit/streams-sd-audiokit.ino), with the Exception, that I use another Constructor for the EncodedAudioStream Objekt (decoder). I simply don't know if I can use it at all, or with another Parameter instead of the AudioKitStream (&i2s) as in the Example.

When I comment out everything after out.begin(cfg); I can see this on the serial Monitor:

`Successfully loaded mp3 File
[I] AudioA2DP.h : 124 - Connecting to JBL GO 2
[I] AudioA2DP.h : 129 - Starting a2dp_source...
[W] AudioA2DP.h : 269 - ==> state: Connecting
[W] AudioA2DP.h : 269 - ==> state: Connected
[  4846][E][BluetoothA2DPSource.cpp:865] bt_app_rc_ct_cb(): [RCCT] Invalid AVRC event: 6`

If I let run the Program any further I get a "Guru Meditation Error":

`Successfully loaded mp3 File
[I] AudioA2DP.h : 124 - Connecting to JBL GO 2
[I] AudioA2DP.h : 129 - Starting a2dp_source...
[W] AudioA2DP.h : 269 - ==> state: Connecting
[W] AudioA2DP.h : 269 - ==> state: Connected
[I] AudioA2DP.h : 143 - a2dp_source is connected...
[I] StreamCopy.h : 74 - buffer_size=1024
Guru Meditation Error: Core  1 panic'ed (LoadProhibited). Exception was unhandled.

Core  1 register dump:
PC      : 0x401a2082  PS      : 0x00060f30  A0      : 0x800d4fe3  A1      : 0x3ffce7d0  
A2      : 0x3ffc3cc4  A3      : 0x00000000  A4      : 0x3f400819  A5      : 0x3ffce948  
A6      : 0x3f41d0ec  A7      : 0xff000000  A8      : 0x3ffb7e4c  A9      : 0x3ffce760  
A10     : 0x00000000  A11     : 0x00000010  A12     : 0x00000010  A13     : 0x3ffce3f0  
A14     : 0x00000010  A15     : 0x3ffce5f0  SAR     : 0x00000014  EXCCAUSE: 0x0000001c  
EXCVADDR: 0x00000000  LBEG    : 0x400905d0  LEND    : 0x400905d5  LCOUNT  : 0x00000000  


Backtrace: 0x401a207f:0x3ffce7d0 0x400d4fe0:0x3ffce7f0 0x400d5132:0x3ffce830 0x400dec15:0x3ffce850`


So the Question would be basically on how I can send whatever I get from the mp3-File to whatever Object I need to send the Stream via Bluetooth. 

I also found the Code from [this Question](https://github.com/pschatzmann/arduino-audio-tools/discussions/475), but unfortunately it also uses the AudioSourceSPIFFS Object.

**Comment by pschatzmann** [comment]
You should not pick the most challenging szenario as your first project: try to learn to walk first before you try to fly!

Try to start with some simple scenarios and test each of your functionality seperatly: otherwise you have no chance to find what you are doing wrong:
https://github.com/pschatzmann/arduino-audio-tools/wiki/It's-not-working

ps. you messed up pretty badly with the audioFile: It should be a global variable but you have it as global which is not used because of your local File audioFile definition. I also recommend to use the basic API using callbacks for this scenario.

**Reply by pdimitri** [reply]
OK, I did my Homework and at least I now can play my mp3 File on my Bluetooth Speaker:

~~~
#include <Arduino.h>
#include <Wire.h>
#include "SPIFFS.h"
#include "AudioTools.h"
#include "AudioLibs/AudioA2DP.h"
#include "AudioCodecs/CodecMP3Helix.h"


A2DPStream out;; // final output of decoded stream
EncodedAudioStream decoder(&out, new MP3DecoderHelix());
File audioFile; 
StreamCopy copier(decoder,audioFile);


void setup(){
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);  


  // setup file
  if(!SPIFFS.begin(true)){
    Serial.println("An Error has occurred while mounting SPIFFS");
    return;
  }

    audioFile = SPIFFS.open("/002.mp3", FILE_READ);
    if(!audioFile){
      Serial.println("Failed to open MP3 file for reading");
      return;
    } else {
      Serial.println("Successfully loaded mp3 File");
    }

    auto cfg = out.defaultConfig(TX_MODE);
    cfg.name = "JBL GO 2";
    out.begin(cfg);

  decoder.setNotifyAudioChange(out);
  decoder.begin();

}

void loop(){
  copier.copy();
}
~~~


But please, one more Question: Even when you say it's easy, I have no Idea on how to play Sounds in response to an Program Event. I don't even know on how to replay the Sound once it had been played once. From my understanding copier.copy() is not a single Event, but more of a Process that needs to be repeated until the whole Stream has been played (and the Buffer has been consumed). That's probably why it's always inside the loop() Function in the Examples. My Program Events most often are catched by Interrupt Functions, that are attached to GPIO Pins and I can't do any  long running or asynchronus Tasks inside these Functions without getting Crashes because of Timeouts. But I could set a Variable in these Functions that later could be used in the loop() Function to specify if and what Sound should be played. But how would I restart the Streaming of the Sound if the conditions are met (i.e Variable set to true)?

**Comment by pschatzmann** [comment]
I think I have already answered this question above, but I try again: The simplest approach would be to move all functionality that you currently call in setup() into a separate method and at the end just call copier.copyAll(): this whis would then be the method that you can call to play a file. Just change the file name into a method parameter and you should have all you need...

If you want to keep the logic with the individual small copy() steps in the loop, you will need to implement some complex state management to figure out in what state the application is.

**Reply by pdimitri** [reply]
Thank you very much for your quick answer and your Patience. But simply copying "all Functionality" from the Setup() Function to another Function would be much too slow in my opinion, as it means to reestablish the Bluetooth Connection and loading the mp3 File from SPIFFS (or SD-Card) every Time I just want to replay a Sound. Therefore I try to figure out wich Parts from the setup() Function I need to redo in my separate Function (play_sound() ) to replay the Sound. This looks good for me, but unfortunately it doesn't work. The Function gets called but no Sound is played. If I would really need to recall  **out.begin(cfg);** it would be useless, because  establishing the BT-Connection to the Speaker envolves a manual Action on the Speaker:

~~~
#include <Arduino.h>
#include <Wire.h>
#include "SPIFFS.h"
#include "AudioTools.h"
#include "AudioLibs/AudioA2DP.h"
#include "AudioCodecs/CodecMP3Helix.h"


A2DPStream out;; // final output of decoded stream
EncodedAudioStream decoder(&out, new MP3DecoderHelix());
File audioFile;
File audioFile2; 
StreamCopy copier(decoder,audioFile);

const int BTN_1 = 34;
bool play_again = false;

void IRAM_ATTR btn_pressed() {
  play_again = true;
}

void play_sound(File file){
  Serial.println("About to play the Sound again");
  StreamCopy copier2(decoder, file);
  decoder.setNotifyAudioChange(out);
  decoder.begin();
  copier2.copyAll();
  play_again = false;
}

void setup(){
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);  

  pinMode(BTN_1, INPUT_PULLUP);
  attachInterrupt(digitalPinToInterrupt(BTN_1), btn_pressed, FALLING);

  // setup file
  if(!SPIFFS.begin(true)){
    Serial.println("An Error has occurred while mounting SPIFFS");
    return;
  }

    audioFile = SPIFFS.open("/002.mp3", FILE_READ);
    if(!audioFile){
      Serial.println("Failed to open MP3 file for reading");
      return;
    } else {
      Serial.println("Successfully loaded mp3 File");
    }

    //audioFile2 = SPIFFS.open("/001.mp3", FILE_READ);

   //decoder.begin();

    auto cfg = out.defaultConfig(TX_MODE);
    cfg.name = "JBL GO 2";
    out.begin(cfg);

  decoder.setNotifyAudioChange(out);
  decoder.begin();

}

void loop(){
  copier.copy();

  if(play_again){
    play_sound(audioFile);
  }
}
~~~

**Comment by pschatzmann** [comment]
Sorry, I overlooked that you want to do this with A2DP: The best way is still, as mentioned, to use basic API. You would just need to make sure that you return an array of 0 if you don't want to send any sound. This way the connection needs to be set up only once. 

I guess this approach should also work using the A2DPStream and writeSilence().

**Reply by pdimitri** [reply]
Looks like my current Problem is the StreamCopy.copyAll() Function. It simply doesn't work for me. Forget about the Program Events for a Moment. I've just commented out everything from the loop() Function and instead used copier.copyAll(); as the last Statement in my Setup Function. Even this doesn't work. The only Thing that gives me at least once an Audio Output is using copier.copy() in the loop()

**Reply by pdimitri** [reply]
Well, it sometimes works. My BT-Speaker makes some Noise when it gets connected to BT, so I sometimes miss the Point when I have to press the "BT Connect" Button on the Speaker during the setup() of the ESP32 and therefore miss the Audio playing. I would really appreciate If there would be any Examples available that show how to "play some Noise when a Button is pressed". This should be really no Rocket Sience, independent of the different Audio Sources and Audio Sinks are used in the Example. Playing Audio makes up 1% of my current Project, but it already took me 40% of the Time invested. Two full Days of Reading, Researching and Testing, just to play a Sound when a Button gets pressed! Playing a Sound on pressing a Button should be nearly as easy as to switch an LED on or off.  

Sorry, I'm just frustrated on how difficult is is to put the Icing on the Cake of my Project.

**Comment by pschatzmann** [comment]
I suggest that you have a look at this example: https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-audiokit/streams-synth-a2dp/streams-synth-a2dp.ino

**Reply by pdimitri** [reply]
Thanks, but have you answered to the right Question? I neither use any MIDI Devices (at least not when it comes to this Project)  nor an AudioKit Device, I just have an Audio Stream that I want to play occationally when certain Conditions are met. If you dont't mind,  just tell me what's wrong with the play_sound() Function from my last Code Example above.

**Reply by pschatzmann** [reply]
This is a demo for showing how to react on simple key pushes to generate a different A2DP output. You asked for an example that show how to "play some Noise when a Button is pressed"

So based on what key is pressed, you would need to make sure that the correct result is returned to the callback method.
